swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll

```diff
@@ -344,11 +344,109 @@ private predicate elementSpec(
   summaryModel(namespace, type, subtypes, name, signature, ext, _, _, _, _)
 }
 
+private string paramsStringPart(AbstractFunctionDecl c, int i) {
+  i = -1 and result = "("
+  or
+  exists(int n, string p | c.getParam(n).getType().toString() = p |
+    i = 2 * n and result = p
+    or
+    i = 2 * n - 1 and result = "," and n != 0
+  )
+  or
+  i = 2 * c.getNumberOfParams() and result = ")"
+}
+
+/**
+ * Gets a parenthesized string containing all parameter types of this callable, separated by a comma.
+ *
+ * Returns the empty string if the callable has no parameters.
+ * Parameter types are represented by their type erasure.
+ */
+cached
+string paramsString(AbstractFunctionDecl c) {
```

Comment:
This part is taken from Java: https://github.com/github/codeql/blob/9cc10e4511992c0c9282cae1187a140c7e0ec5d5/java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll#L657